rfcs/001-mid-level-file-api.md

```diff
@@ -0,0 +1,187 @@
+# Mid-Level File Api
+
+## Summary
+
+This is a mid-level API wrapper around the various File related JavaScript apis. The possibility for a higher level API is left open.
+
+The mid-level file API aims to implement File I/O on top of the raw JS apis found here: https://w3c.github.io/FileAPI/ which includes listing files, creating files, and reading files.
+
+## The API
+
+First we have `FileList` which is simply a list of `File`s. This is an opaque struct that offers a way to iterate over `Files` and get individual files by index:
```

Comment:
This would require iterating the internal `js_sys::FileList` and collecting the files into a `Vec`. Are we sure we want the user to pay this price each time a `FileList` is constructed? I personally think an `into_vec` API is more appropriate since it conveys that there will be an additional allocation.

In any effect, it's possible to add these in a backward compatible way. Can we push this off to another PR/RFC?

Reply:
I don't think it's possible to add it in a backwards compatible way, since a `Vec` would have a different API.

And since (apparently) `FileList` is being phased out in favor of a normal JS `Array`, I think `Vec` is quite appropriate.

In fact, we may even want to skip the `FileList` type entirely and just return `Vec<File>`.

That also makes it pretty clear that there is an allocation (since `Vec` obviously allocates).

Reply:
@Pauan Can you explain how the API would need to be different? I'm sorry, I don't really get the reason for this given the cost users would have to pay.

Reply:
Basically, rather than having a method which returns a `FileList`, it would instead return `Vec<File>`.

That means we don't actually need `FileList` at all.

The cost is quite small, but the benefits are significant: things are drastically simplified, and users get to use the slice methods for free.